[WFLY-12457] Single action timer is not triggered automatically after a DB outage, requires server restart

elguardian opened the pull request:
added logic when in memory timer are out of sync with the store

https://issues.jboss.org/browse/WFLY-12457

bstansberry commented:
Hi @elguardian  -- Please file a WFLY JIRA for this and use it in the commit message and PR title. Thanks!

Commit:
[WFLY-12457] Single action timer is not triggered automatically after a DB outage, requires server restart

when the timer in memory and in database has different states
in memory : end state
store : still active
means that the timer could not be executed causing an inconsistency. The only way to retrigger it was to
restart the server. this detects and acts acordingly

elguardian commented:
@bstansberry done, who should I ask for review ?
Cannot access the jdk 11 errors.

bstansberry commented:
@chengfang or @tadamski Please review when you get a chance.

@elguardian The JDK 11 errors look unrelated.

chengfang commented on ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/TimerPersistence.java:
```diff
@@ -96,6 +96,12 @@
+        /**
+         * Invoked when a timer needs to be sync with the underlying store
+         * @param timer The timer
+         */
```
document both params.

elguardian replied:
done

chengfang commented on ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java:
```diff
@@ -839,6 +840,23 @@ public void run() {
+                                        boolean validDBTimer = valid.contains(holder.timer.getState());
+                                        boolean validMemoryTimer = !valid.contains(oldTimer.getState());
+                                        // if timers memory - db are in non intersect substes of valid/invalid states. we put them in sync
+                                        if (holder != null && validMemoryTimer && validDBTimer) {
```
move holder != null check earlier before its first access.

elguardian replied:
done

chengfang reviewed (commented):
Looks good other than a few minor points noted inline.

elguardian commented:
@chengfang both checks are done. let me know if there is anything to make this move forward

bstansberry committed:
[WFLY-12457] Remove trailing space

Fix checkstyle failure

bstansberry commented:
@elguardian I added a commit fixing the checkstyle problem.

elguardian commented:
thx @bstansberry
